bump spack/ubuntu-jammy wave build base image to `develop`

munishchouhan opened the pull request:
This PR will bump ubuntu-jammy to 0.21.0
it could help to resolve https://github.com/seqeralabs/wave/issues/351

pditommaso committed:
[release] bump version 1.1.3

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Fix NPE when accessing pollDuration

The pollDuration field in the PairingOutboundQueue class
must be declared final, otherwise when it's accessed
by the polling thread a null value can be returned.

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Minor change on UnexpectedReadException error message

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Use virtual threads for I/O bound operations

This commit introduces the use of virtual threads for short-lived
I/O bound operations. More in details virtual threads are used in
place of the usual thread pool for:
- PairingInboundStore: use a virtual thread pool to creare a CompletableFuture
  to await the completation of a remote command
- HttpClientFactory: use a virtual thread for HttpClient operations
- TimedInputStream: use a virtual thread pool to read the target input
  stream within a fixed amount of time.

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Use volatile to prevent invalid mem access

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
[release] bump version 1.1.4

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Revert "Fix Container 302 redirect response with body"

This reverts commit 1a90cbc2f5f1f855fa8b6ea8c937cb01538c96b4.

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Fix Container 302 redirect response with body (2nd version)

This commit fixes the handling of redirect response made by
some registry e.g. Quay.io which can provide a repsonse body
other than the Location header directive.

The patch consists of just overring the Content-Length header
to zero, instead of using the length returned by the original
request and reading the response body as done in the commit
1a90cbc2.

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Add retry policy when reading manifest response

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
[release] bump version 1.1.5

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Rollback virtual threads for TimedInputStream read

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Bump trivy 0.47.0

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

munishchouhan committed:
Refactor build settings via BuildConfig class (#324)

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Co-authored-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

jason-seqera committed:
[docs] Simplify sidebar slightly (#350)

This rolls up the CLI index into the category itself. This simplifies
the sidebar and uses space more efficiently.

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

jason-seqera committed:
[docs] Add visibility into API rate limits (#348)

This content is also included already for the CLI.

- https://github.com/seqeralabs/nf-tower-docs/issues/477

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

jason-seqera committed:
[docs] Remove GitHub workflows for legacy docs site (#349)

This no longer builds the public documentation.

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

jason-seqera committed:
[docs] Use cr.seqera.io/public/docs-previewer image (#352)

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Update changelog

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@seqera.io>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

munishchouhan committed:
bump ubuntu-jammy to 0.21.0

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

bebosudo commented:
This should be solved once we implement the architecture detailed in https://github.com/seqeralabs/devops-backlog/issues/269.

marcodelapierre commented:
for reference, we have that version here as well: `https://github.com/seqeralabs/wave/blob/41877a928bf469eafa6286fa6e751dce909c3a85/src/main/resources/io/seqera/wave/spack/spack-builder-dockerfile.txt#L20-L27`

marcodelapierre commented:
Hi @munishchouhan  please bump these 3 occurrences to `develop`, as per our chat with @bebosudo 👍 
```
$ grep -r 'v0.20'  src/
src//main/resources/io/seqera/wave/spack/spack-builder-singularityfile.txt:    spack mirror add binary_mirror  https://binaries.spack.io/releases/v0.20
src//main/resources/io/seqera/wave/spack/spack-builder-dockerfile.txt:&& spack mirror add binary_mirror  https://binaries.spack.io/releases/v0.20 \
src//main/groovy/io/seqera/wave/configuration/SpackConfig.groovy:    @Value('${wave.build.spack.builderImage:`spack/ubuntu-jammy:v0.20.0`}')
```

marcodelapierre commented:
As for the user's GH issue, note that Spack in Wave is currently in Beta, as mentioned in the docs: https://www.nextflow.io/docs/latest/wave.html#build-spack-based-containers.

We can mention them that we are currently working towards increased stability, robustness and usefulness of this functionality, for future releases.

pditommaso committed:
Remove unsused consts

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

marcodelapierre committed:
spack binary cache and base image : using develop version

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@seqera.io>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

marcodelapierre commented:
@munishchouhan  can you take a quick look please?

There was a non-signed / non verified commit. I tried and fix that, with no success. So we still need to verify the commits in this PR.

marcodelapierre commented:
I also don't get why the CI is now failing

pditommaso commented:
what if a user wants to use a tool in version 0.20.0?

marcodelapierre commented:
Also, see this nextflow PR: we could actually update the 3 fields to use the Spack version as provided by NF:

https://github.com/nextflow-io/nextflow/pull/4554

(at this stage, I would call for the default in NF to be `develop`)

munishchouhan committed:
failed UT fixed

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

munishchouhan commented:
> I also don't get why the CI is now failing

@marcodelapierre fixed the errors

marcodelapierre commented:
@bebosudo ready for your review

(after that, @munishchouhan  you will still need to add Verification to the commits, for the merge to be actionable)

bebosudo reviewed (approved):
LGTM, thanks

marcodelapierre commented:
@pditommaso ready to be merged 
* you turned it into draft, so I won't undo it without your green lights
* please consider my point of discussing `develop` vs other versions out of this scope ; this will enable Alberto and the team to go ahead with testing the setup proposed in https://github.com/seqeralabs/devops-backlog/issues/269

marcodelapierre commented:
@pditommaso please action this, eventually after Wave meeting on 19 December, where you can discuss with @bebosudo

pditommaso commented:
There's no pressure. It's broken anyway

pditommaso committed:
Fix typo in comment

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Update failing tests

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Implements streaming client for large binary downloads (#358)

This commit implements the use of ReactorStreamingHttpClient
to stream the layer blob binaries, in place of the Java SDK HTTP client
which shows several memory issues for large payloads.

Moreover, it uses the setting wave.httpclient.streamThreshold (default 64k)
to determine when to read the layer blob directly into a byte[] buffer or
use a separate HTTP request and stream the content for large blobs

---------

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Improve usage of thread pools for build and scan service

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
[release] bump version 1.1.7

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Update changelog

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Bump kaniko 1.19.2

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

munishchouhan committed:
Fix annotations processing package (#364)

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

bebosudo commented:
Could you comment on this Paolo? What's still broken?

bebosudo commented on src/main/groovy/io/seqera/wave/configuration/SpackConfig.groovy:
```diff
@@ -61,7 +61,7 @@ class SpackConfig {
      * The container image used for Spack builds
      */
-    @Value('${wave.build.spack.builderImage:`spack/ubuntu-jammy:v0.20.0`}')
+    @Value('${wave.build.spack.builderImage:`spack/ubuntu-jammy:develop`}')
```
```suggestion
    @Value('${wave.build.spack.builderImage:`cr.seqera.io/spack/wave/spack/ubuntu-jammy:develop`}')
```

I actually just realized that moving to `spack/ubuntu-jammy:develop` won't be useful, since our workflow goes like this:
- a cron job starts every day to build spack packages in our spack stack against `spack/ubuntu-jammy:develop`
- the built packages are pushed to our build cache
- only if the build succeeded, retag `spack/ubuntu-jammy:develop` into `195996028523.dkr.ecr.eu-west-1.amazonaws.com/wave/spack/ubuntu-jammy:develop`, which is proxied by Harbor at the address in the suggestion above

So we should make wave use our `develop` image, not the one from the spack project, because that may break for our spack stack, and the GHA would notify us of that.

Here's the cron job I built to test the GHA that automatically performs the actions above: https://github.com/seqeralabs/test-nf-tower-cloud-gha-dev-branch-pr/blob/master/.github/workflows/populate-spack-binary-mirror.yml

marcodelapierre committed:
Update src/main/groovy/io/seqera/wave/configuration/SpackConfig.groovy

Co-authored-by: Alberto Chiusole <1922124+bebosudo@users.noreply.github.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

marcodelapierre committed:
Revert "Update src/main/groovy/io/seqera/wave/configuration/SpackConfig.groovy"

This reverts commit 06e6e2dd9dbf266a2a6752dc01bbb0b6f92e8814.

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

marcodelapierre committed:
Reapply "Update src/main/groovy/io/seqera/wave/configuration/SpackConfig.groovy"

This reverts commit 2d684c01c1f2c3cfbe819769d7b36423a2495017.

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

jason-seqera committed:
Copy back copy edits (#365)

* Copy back copy edits

- https://github.com/seqeralabs/docs/pull/48

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@seqera.io>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

* Minor style edit in guide.md

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@seqera.io>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

---------

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@seqera.io>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>
Co-authored-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

munishchouhan committed:
Changed wave.build.public to wave.build.public-repo (#372)

* added repo to wave.build.public

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

* changed to publicRepo

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

* changed to kebab case

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

---------

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Remove asciidocker gradle task

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

marcodelapierre committed:
build ci: accept slash in branch name (#377)

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Add support for blob caching (#359)

This commit adds the ability to cache into an S3-compatible bucket
container layer blobs to prevent multiple downloads from the remote
registry of the same multiple times. Moreover, it offloads the serving of
the blob payloads to the object storage, creating a pre-signed HTTP URL
for the layer blobs to which the Docker client is redirected.

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: munishchouhan <hrma017@gmail.com>
Co-authored-by: Munish Chouhan <hrma017@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Fix bootstrap error when blobCache is disabled

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Fix typo in comment [ci skip]

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
Implement support for Cloudflare WAF token  (#383)

This PR adds support for Cloudflare WAF authenticating token.

This allows caching the image blob layers into a R2 bucket and serve
them via Cloudflare CDN restricting the access only to request having
the `verify` token in the request URL, i.e. preventing the public access
to those objects.

This feature requires providing these configuration settings

```
wave:
      blobCache:
        enabled: true
        baseUrl: "<CDN base URL e.g. https://bucket-name.com>"
        signing-strategy: "cloudflare-waf-token"
        cloudflare:
          lifetime: "<token ttl e.g. 3h>"
          secret-key: "<your waf secret>"
```

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

pditommaso committed:
[release] bump version 1.2.0

Signed-off-by: Paolo Di Tommaso <paolo.ditommaso@gmail.com>
Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

marcodelapierre committed:
wave spack: runnerImage renamed from 22.04 to jammy, same sha, to be consistent with builderImage

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

marcodelapierre commented:
@pditommaso  @bebosudo  merging as we agreed on the use of `develop` in the Spack infra

marcodelapierre committed:
Merge branch 'master' into bump-ubuntu-jammy-to-0-21-0

Signed-off-by: Dr Marco Claudio De La Pierre <marco.delapierre@gmail.com>

marcodelapierre commented:
Closing in favour of https://github.com/seqeralabs/wave/pull/385/commits due to commit signing off issues
